# ADS 502 Moduel 4 Assignment  | Python 

### Ryan S. Dunn 

## Data Science Using Python and R: Chapter 13 - Page 195: Questions #13, 14, 15, 16, & 17

### 13. Create a logistic regression model to predict whether or not a customer has a store credit card, based on whether they have a web account and the days between purchases. Obtain the summary of the model. 

In [90]:
#import libraries
import pandas as pd
import numpy as np 
import statsmodels.api as sm
from scipy import stats
from sklearn.metrics import confusion_matrix

In [2]:
clothing_test = pd.read_csv("/Users/ryan_s_dunn/Documents/USD MS-ADS/Applied Data Mining 502/Module 4/Datasets/clothing_sales_test.csv")

In [3]:
clothing_train = pd.read_csv("/Users/ryan_s_dunn/Documents/USD MS-ADS/Applied Data Mining 502/Module 4/Datasets/clothing_sales_training.csv")

In [4]:
#separate the training variables into predictor variables and response variables
X_train = pd.DataFrame(clothing_train[['Days','Web']])
X_train = sm.add_constant(X_train)
y_train = pd.DataFrame(clothing_train[['CC']])

In [5]:
#develop the logistic regression model 
logreg1 = sm.Logit(y_train, X_train).fit()
logreg1.summary2()

Optimization terminated successfully.
         Current function value: 0.655955
         Iterations 5


<class 'statsmodels.iolib.summary2.Summary'>
"""
                         Results: Logit
=================================================================
Model:              Logit            Pseudo R-squared: 0.053     
Dependent Variable: CC               AIC:              1909.5825 
Date:               2021-11-21 18:42 BIC:              1925.4226 
No. Observations:   1451             Log-Likelihood:   -951.79   
Df Model:           2                LL-Null:          -1004.9   
Df Residuals:       1448             LLR p-value:      8.3668e-24
Converged:          1.0000           Scale:            1.0000    
No. Iterations:     5.0000                                       
-------------------------------------------------------------------
           Coef.    Std.Err.      z      P>|z|     [0.025    0.975]
-------------------------------------------------------------------
const      0.4962     0.0887    5.5968   0.0000    0.3224    0.6699
Days      -0.0037     0.0004   -8.4491   0.0000   -0.0046   -0.0028
Web        1.2537     0.3307    3.7914   0.0001    0.6056    1.9018
=================================================================

"""

### 14. Are there any variables that should be removed from the model? If so, remove them and rerun the model

There should not be any variables removed from the model, as both p-values for Days and Web suggest that the co-efficients are statistically significant, and therefore relevant for the logistic regression model.

### 15. Write the descriptive form of the logistic regression model using the coefficients obtained from Question #13. 

$\ CC = \beta_0 - \beta_1(Days) - \beta_2(Web)$

Where: 
$\beta_0 = 0.4962$ ,
$\beta_1 = 0.0037$ ,
$\beta_2 = 1.2537$

### 16. Validate the model using the test data set.

In [6]:
#separate the test variables into predictor variables and response variables
X_test = pd.DataFrame(clothing_test[['Days','Web']])
X_test = sm.add_constant(X_test)
y_test = pd.DataFrame(clothing_test[['CC']])

#run the model using the test data set
logreg1_test = sm.Logit(y_test, X_test).fit()
logreg1_test.summary2()

Optimization terminated successfully.
         Current function value: 0.656885
         Iterations 5


<class 'statsmodels.iolib.summary2.Summary'>
"""
                         Results: Logit
=================================================================
Model:              Logit            Pseudo R-squared: 0.052     
Dependent Variable: CC               AIC:              1838.7104 
Date:               2021-11-21 18:42 BIC:              1854.4324 
No. Observations:   1395             Log-Likelihood:   -916.36   
Df Model:           2                LL-Null:          -966.40   
Df Residuals:       1392             LLR p-value:      1.8534e-22
Converged:          1.0000           Scale:            1.0000    
No. Iterations:     5.0000                                       
-------------------------------------------------------------------
           Coef.    Std.Err.      z      P>|z|     [0.025    0.975]
-------------------------------------------------------------------
const      0.4634     0.0873    5.3105   0.0000    0.2924    0.6345
Days      -0.0035     0.0004   -8.2261   0.0000   -0.0043   -0.0026
Web        1.0973     0.2830    3.8780   0.0001    0.5427    1.6519
=================================================================

"""

### 17. Obtain the predicted values of the response variable for each record in the data set. 

In [91]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

#call the LogisticReggresion algorithm from sklearn
logreg = LogisticRegression()
logreg.fit(X_train, y_train)

#obtain the predicted values from the test set
y_pred = logreg.predict(X_test)

#create the "predicted and actual" confusion matrix of the actual and predicted values 
conf_max_lr = pd.DataFrame(confusion_matrix(y_test, y_pred))
conf_max_lr

/Users/ryan_s_dunn/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


0  1
0  23886  0
1   2988  0

In [8]:
#obtain the classification report for the training data
#print(classification_report(y_test, y_pred))

## Data Science Using Python and R: Chapter 9 - Page 138: Questions #24, 25, 26, 27, 28, 29, & 30

#### 24. Prepare the data set for neural network modeling, including standardizing the variables.

In [39]:
import pandas as pd
import numpy as np

#import training and test data sets
bank_test = pd.read_csv("/Users/ryan_s_dunn/Documents/USD MS-ADS/Applied Data Mining 502/Module 4/Datasets/bank_marketing_test")

#import the test data set 
bank_train = pd.read_csv("/Users/ryan_s_dunn/Documents/USD MS-ADS/Applied Data Mining 502/Module 4/Datasets/bank_marketing_training")

In [10]:
#preprocess the data for the ANN model
bank_train.head()

age          job  marital    education  default housing loan    contact  \
0   56    housemaid  married     basic.4y       no      no   no  telephone   
1   57     services  married  high.school  unknown      no   no  telephone   
2   41  blue-collar  married      unknown  unknown      no   no  telephone   
3   25     services   single  high.school       no     yes   no  telephone   
4   29  blue-collar   single  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  days_since_previous  previous  \
0   may         mon  ...         1                  999         0   
1   may         mon  ...         1                  999         0   
2   may         mon  ...         1                  999         0   
3   may         mon  ...         1                  999         0   
4   may         mon  ...         1                  999         0   

   previous_outcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0       nonexistent          1.1          93.994          -36.4      4.857   
1       nonexistent          1.1          93.994          -36.4      4.857   
2       nonexistent          1.1          93.994          -36.4      4.857   
3       nonexistent          1.1          93.994          -36.4      4.857   
4       nonexistent          1.1          93.994          -36.4      4.857   

   nr.employed  response  
0         5191        no  
1         5191        no  
2         5191        no  
3         5191        no  
4         5191        no  

[5 rows x 21 columns]

In [11]:
#develop the ANN model
#import tensorflow as tf

### 29. Using the same predictors you used for your neural network model, buidl models to predict Response using the following algorithms:

### a. CART

In [94]:
#include neccessary packages to create and display the CART model
import statsmodels.tools.tools as stattools
from sklearn.tree import DecisionTreeClassifier, export_graphviz 
from sklearn import tree

y_train = bank_train[['response']]
y_test = bank_test[['response']]

In [95]:
marital_dummy = np.array(bank_train['marital'])
education_dummy = np.array(bank_train['education'])
loan_dummy = np.array(bank_train['loan'])

(marital_cat, martical_cat_dict) = stattools.categorical(marital_dummy, drop=True, dictnames=True)
(education_cat, education_cat_dict) = stattools.categorical(education_dummy, drop=True, dictnames=True)
(loan_cat, loan_cat_dict) = stattools.categorical(loan_dummy, drop=True, dictnames=True)

marital_status = pd.DataFrame(marital_cat)
education_status = pd.DataFrame(education_cat)
loan_status = pd.DataFrame(loan_cat)

#develop the predictor variable training dataframe 
X_train = pd.concat((marital_status, education_status, loan_status), axis =1)

/Users/ryan_s_dunn/opt/anaconda3/lib/python3.8/site-packages/statsmodels/tools/tools.py:158: FutureWarning: categorical is deprecated. Use pandas Categorical to represent categorical data and can get_dummies to construct dummy arrays. It will be removed after release 0.13.
  warnings.warn(


In [96]:
#prep the test data
marital_dummy_test = np.array(bank_test['marital'])
education_dummy_test = np.array(bank_test['education'])
loan_dummy_test = np.array(bank_test['loan'])

(marital_cat_test, martical_cat_dict_test) = stattools.categorical(marital_dummy_test, drop=True, dictnames=True)
(education_cat_test, education_cat_dict_test) = stattools.categorical(education_dummy_test, drop=True, dictnames=True)
(loan_cat_test, loan_cat_dict_test) = stattools.categorical(loan_dummy_test, drop=True, dictnames=True)

marital_status_test = pd.DataFrame(marital_cat_test)
education_status_test = pd.DataFrame(education_cat_test)
loan_status_test = pd.DataFrame(loan_cat_test)

#develop the test predictor varialbe dataframe 
X_test = pd.concat((marital_status_test, education_status_test, loan_status_test), axis =1)

/Users/ryan_s_dunn/opt/anaconda3/lib/python3.8/site-packages/statsmodels/tools/tools.py:158: FutureWarning: categorical is deprecated. Use pandas Categorical to represent categorical data and can get_dummies to construct dummy arrays. It will be removed after release 0.13.
  warnings.warn(


In [104]:
#develop the CART model with the training data
cart = DecisionTreeClassifier(criterion = 'gini', max_leaf_nodes =5).fit(X_train, y_train)

In [105]:
#get the predicted values for X_test with the CART model 
y_pred_cart = cart.predict(X_test)

In [106]:
#obtain the classification report for the training data
print(classification_report(y_pred_cart, y_test))

/Users/ryan_s_dunn/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

          no       1.00      0.89      0.94     26874
         yes       0.00      0.00      0.00         0

    accuracy                           0.89     26874
   macro avg       0.50      0.44      0.47     26874
weighted avg       1.00      0.89      0.94     26874



/Users/ryan_s_dunn/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/ryan_s_dunn/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [107]:
#confusion matrix for CART
conf_max_cart = pd.DataFrame(confusion_matrix(y_test, y_pred_cart))
conf_max_cart

0  1
0  23886  0
1   2988  0

### b. C5.0

In [82]:
#develop the c5 model
c5 = DecisionTreeClassifier(criterion="entropy", min_samples_split=75, max_leaf_nodes=5).fit(X_train, y_train)

In [86]:
#run the C5 test data in the model 
y_pred_c5 = c5.predict(X_test)

In [87]:
#obtain the classification report for the training data
print(classification_report(y_pred_c5, y_test))

/Users/ryan_s_dunn/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

          no       1.00      0.89      0.94     26874
         yes       0.00      0.00      0.00         0

    accuracy                           0.89     26874
   macro avg       0.50      0.44      0.47     26874
weighted avg       1.00      0.89      0.94     26874



/Users/ryan_s_dunn/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/ryan_s_dunn/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [102]:
#confusion matrix for C5
conf_max_c5 = pd.DataFrame(confusion_matrix(y_test, y_pred_c5))
conf_max_c5

0  1
0  23886  0
1   2988  0

### c. Naive Bayes

In [13]:
import pandas as pd
import numpy as np 
from sklearn.naive_bayes import MultinomialNB
import statsmodels.tools.tools as statstools

In [14]:
#X and Y variables for the Naive Bayes model 
X_marital_ind = np.array(bank_train['marital'])
(X_marital_ind, X_marital_ind_dict) = statstools.categorical(X_marital_ind, drop = True, dictnames = True)
X_marital_ind = pd.DataFrame(X_marital_ind)

X_education_ind = np.array(bank_train['education'])
(X_education_ind, X_education_ind_dict) = statstools.categorical(X_education_ind, drop = True, dictnames = True)
X_education_ind = pd.DataFrame(X_education_ind)

X_loan_ind = np.array(bank_train['loan'])
(X_loan_ind, X_loan_ind_dict) = statstools.categorical(X_loan_ind, drop = True, dictnames = True)
X_loan_ind = pd.DataFrame(X_loan_ind)

X = pd.concat((X_marital_ind, X_education_ind, X_loan_ind), axis = 1)

y_response = bank_train['response']

In [15]:
#develop the Naive Bayes model
nb_bank = MultinomialNB().fit(X,y_response)

In [16]:
#test data preprocessing
X_marital_ind_test = np.array(bank_test['marital'])
(X_marital_ind_test, X_marital_ind_dict_test) = statstools.categorical(X_marital_ind_test, drop = True, dictnames = True)
X_marital_ind_test = pd.DataFrame(X_marital_ind_test)

X_education_ind_test = np.array(bank_test['education'])
(X_education_ind_test, X_education_ind_dict_test) = statstools.categorical(X_education_ind_test, drop = True, dictnames = True)
X_education_ind_test = pd.DataFrame(X_education_ind_test)

X_loan_ind_test = np.array(bank_test['loan'])
(X_loan_ind_test, X_loan_ind_dict_test) = statstools.categorical(X_loan_ind_test, drop = True, dictnames = True)
X_loan_ind_test = pd.DataFrame(X_loan_ind_test)

X_test = pd.concat((X_marital_ind_test, X_education_ind_test, X_loan_ind_test), axis = 1)
y_test = bank_test['response']

/Users/ryan_s_dunn/opt/anaconda3/lib/python3.8/site-packages/statsmodels/tools/tools.py:158: FutureWarning: categorical is deprecated. Use pandas Categorical to represent categorical data and can get_dummies to construct dummy arrays. It will be removed after release 0.13.
  warnings.warn(


In [17]:
#run the model against the test data 
y_predicted = nb_bank.predict(X_test)

In [18]:
ypred_bank = pd.crosstab(bank_test['response'], y_predicted, rownames = ['Actual'], colnames = ['Predicted'])
ypred_bank['Total'] = ypred_bank.sum(axis=1); ypred_bank.loc['Total'] = ypred_bank.sum(); 
ypred_bank

Predicted     no  Total
Actual                 
no         23886  23886
yes         2988   2988
Total      26874  26874

In [19]:
from sklearn.metrics import classification_report

print(classification_report(y_predicted, y_test))

/Users/ryan_s_dunn/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

          no       1.00      0.89      0.94     26874
         yes       0.00      0.00      0.00         0

    accuracy                           0.89     26874
   macro avg       0.50      0.44      0.47     26874
weighted avg       1.00      0.89      0.94     26874



/Users/ryan_s_dunn/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/ryan_s_dunn/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## Data Science Using Python and R: Chapter 6 - Page 93: Questions #19 & 20

In [130]:
#preprocess the data for inclusion in the random forest algorithm
import statsmodels.tools.tools as stattools
from sklearn.tree import DecisionTreeClassifier, export_graphviz 
from sklearn import tree

In [131]:
#import training data
training = pd.read_csv("/Users/ryan_s_dunn/Documents/USD MS-ADS/Applied Data Mining 502/Module 2/Datasets/adult_ch6_training", header = 0)

In [132]:
#create our y value - Income
y_train_rf = training['Income']

In [133]:
#convert categorical varialbes to dummy variables because sklearn needs converted categorical variables
marital_status_dummy = np.array(training['Marital status'])

#save matrix and dictionary separately
(marital_cat, marital_cat_dict) = stattools.categorical(marital_status_dummy, 
                                                        drop=True,dictnames=True)

/Users/ryan_s_dunn/opt/anaconda3/lib/python3.8/site-packages/statsmodels/tools/tools.py:158: FutureWarning: categorical is deprecated. Use pandas Categorical to represent categorical data and can get_dummies to construct dummy arrays. It will be removed after release 0.13.
  warnings.warn(


In [134]:
#add the dummy variables back into the X varialbes
marital_status_pd = pd.DataFrame(marital_cat)

#attach the predictor variable Cap_Gains_Losses to the data frame of dummy variables that represent marital status
X_train= pd.concat((training[['Cap_Gains_Losses']],
                    marital_status_pd), axis=1)

In [135]:
#note that the columns of X and y do not include the different values of the marital status, so we run 
#martial_cat_dict and assigne the categories to y to see the values of the columns
X_names = ["Cap_Gains_Losses","Divorced","Married",
           "Never-married","Separated","Widowed"]
y_names = ["<=50K",">50K"]

### 19. Use random forests on the training data set to predict income using marital status and capital gains and losses. 

In [136]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
import numpy as np 

In [137]:
#include the response variable formattted as a one dimensional array
rfy = np.ravel(y_train_rf)

In [138]:
#use the randomforest classifier to create the randome forest
rf_1 = RandomForestClassifier(n_estimators = 100, criterion ="gini").fit(X_train, rfy)    

In [139]:
#obtain the predictions from the model 
yrf_predict = rf_1.predict(X_train)

#return the confusion matrix for the random forest classifier
cm_yrf = pd.DataFrame(confusion_matrix(y_train, yrf_predict))
cm_yrf

0     1
0  14237    34
1   3138  1352

In [140]:
#obtain the classification report for the training data
print(classification_report(y_train, yrf_predict))

              precision    recall  f1-score   support

       <=50K       0.82      1.00      0.90     14271
        >50K       0.98      0.30      0.46      4490

    accuracy                           0.83     18761
   macro avg       0.90      0.65      0.68     18761
weighted avg       0.86      0.83      0.79     18761



### 20. Use random forests using the test data set that utilizes the same target and predictor variables. Does the test data result match the training data result?

In [141]:
#import test data
test = pd.read_csv("/Users/ryan_s_dunn/Documents/USD MS-ADS/Applied Data Mining 502/Module 2/Datasets/adult_ch6_test", header = 0)

In [142]:
#identify the target variable in the test data set 
y_test = test['Income']

#convert categorical varialbes to dummy variables because sklearn needs converted categorical variables
marital_status_dummy2 = np.array(test['Marital status'])

#save matrix and dictionary separately
(marital_cat2, marital_cat_dict2) = stattools.categorical(marital_status_dummy2, 
                                                        drop=True,dictnames=True)

#add the dummy variables back into the X varialbes
marital_status_pd2 = pd.DataFrame(marital_cat2)

#attach the predictor variable Cap_Gains_Losses to the data frame of dummy variables that represent marital status
X_test= pd.concat((test[['Cap_Gains_Losses']],
                    marital_status_pd2), axis=1)

#note that the columns of X and y do not include the different values of the marital status, so we run 
#martial_cat_dict and assigne the categories to y to see the values of the columns
X_names2 = ["Cap_Gains_Losses","Divorced","Married",
           "Never-married","Separated","Widowed"]
y_names2 = ["<=50K",">50K"]

/Users/ryan_s_dunn/opt/anaconda3/lib/python3.8/site-packages/statsmodels/tools/tools.py:158: FutureWarning: categorical is deprecated. Use pandas Categorical to represent categorical data and can get_dummies to construct dummy arrays. It will be removed after release 0.13.
  warnings.warn(


In [143]:
from sklearn.metrics import plot_confusion_matrix
from sklearn.metrics import ConfusionMatrixDisplay

#include the response variable formattted as a one dimensional array
rfy2 = np.ravel(y_test)

#use the randomforest classifier to create the randome forest
rf_2 = RandomForestClassifier(n_estimators = 100, criterion ="gini").fit(X_test,rfy2)  

#obtain the predictions from the model 
yrf_predict_test = rf_2.predict(X_test)

#return the confusion matrix for the random forest classifier
cm_yrf = pd.DataFrame(confusion_matrix(y_test, yrf_predict_test))
print('Random Forest Confusion Matrix:')
print(cm_yrf)


Random Forest Confusion Matrix:
      0    1
0  4668    6
1  1034  447


In [144]:
#obtain the classification report for the test data
print(classification_report(y_test, yrf_predict_test))

              precision    recall  f1-score   support

       <=50K       0.82      1.00      0.90      4674
        >50K       0.99      0.30      0.46      1481

    accuracy                           0.83      6155
   macro avg       0.90      0.65      0.68      6155
weighted avg       0.86      0.83      0.79      6155



In [145]:
#obtain the classification report for the training data
print(classification_report(y_train, yrf_predict))

              precision    recall  f1-score   support

       <=50K       0.82      1.00      0.90     14271
        >50K       0.98      0.30      0.46      4490

    accuracy                           0.83     18761
   macro avg       0.90      0.65      0.68     18761
weighted avg       0.86      0.83      0.79     18761



##### The test training result does match the test training results. 